### PR TITLE
feat: New Module `avm/res/machine-learning-services/workspace`

### PR DESCRIPTION
**Pull request by krbar**
## Description

New Module `avm/res/machine-learning-services/workspace`, migrated from CARML.

## Adding a new module

- [x] A proposal has been submitted and approved.
- [ ] I have included "Closes #{module_proposal_issue_number}" in the PR description.
- [ ] I have run `brm validate` locally to verify the module files.
- [x] I have run deployment tests locally to ensure the module is deployable.

## Pipeline references

| Pipeline |
| - |
| [avm.res.machine-learning-services.workspace](https://github.com/krbar/bicep-registry-modules/actions/workflows/avm.res.machine-learning-services.workspace.yml) |

### CODE REVIEW

File: avm/res/machine-learning-services/workspace/compute/main.bicep [hunk 1]
```diff
@@ -0,0 +1,144 @@
+metadata name = 'Machine Learning Services Workspaces Computes'
```
Comment (AlexanderSehr) [inline avm/res/machine-learning-services/workspace/compute/main.bicep:1]:
```suggestion
metadata name = 'Machine Learning Services Workspaces'
```

File: avm/res/machine-learning-services/workspace/main.bicep [hunk 2]
```diff
@@ -0,0 +1,477 @@
...
+@sys.description('Optional. The lock settings of the service.')
+param lock lockType
+
+@sys.description('Optional. The flag to signal HBI data in the workspace and reduce diagnostic data collected by the service.')
+param hbiWorkspace bool = false
+
+@sys.description('Optional. The flag to indicate whether to allow public access when behind VNet.')
+param allowPublicAccessWhenBehindVnet bool = false
+
+@sys.description('Optional. Array of role assignment objects that contain the \'roleDefinitionIdOrName\' and \'principalId\' to define RBAC role assignments on this resource. In the roleDefinitionIdOrName attribute, you can provide either the display name of the role definition, or its fully qualified ID in the following format: \'/providers/Microsoft.Authorization/roleDefinitions/c2f4ef07-c644-48eb-af81-4b1b4947fb11\'.')
```
Comment (AlexanderSehr) [inline avm/res/machine-learning-services/workspace/main.bicep:44]:
```suggestion
@sys.description('Optional. Array of role assignments to create.')
```

File: avm/res/machine-learning-services/workspace/main.bicep [hunk 3]
```diff
@@ -0,0 +1,477 @@
...
+output resourceId string = workspace.id
+
+@sys.description('The resource group the machine learning service was deployed into.')
+output resourceGroupName string = resourceGroup().name
+
+@sys.description('The name of the machine learning service.')
+output name string = workspace.name
+
+@sys.description('The principal ID of the system assigned identity.')
+output systemAssignedMIPrincipalId string = (managedIdentities.?systemAssigned ?? false) && contains(workspace.identity, 'principalId') ? workspace.identity.principalId : ''
```
Comment (AlexanderSehr) [inline avm/res/machine-learning-services/workspace/main.bicep:305]:
```suggestion
output systemAssignedMIPrincipalId string = workspace.?identity.?principalId ?? ''
```

File: avm/res/machine-learning-services/workspace/main.bicep [hunk 4]
```diff
@@ -0,0 +1,477 @@
...
+      keyVaultProperties: {
+        keyVaultArmId: cMKKeyVault.id
+        keyIdentifier: !empty(customerManagedKey.?keyVersion ?? '') ? '${cMKKeyVault::cMKKey.properties.keyUri}/${customerManagedKey!.keyVersion}' : cMKKeyVault::cMKKey.properties.keyUriWithVersion
+      }
+    } : null
+    imageBuildCompute: imageBuildCompute
+    primaryUserAssignedIdentity: primaryUserAssignedIdentity
+    publicNetworkAccess: !empty(publicNetworkAccess) ? any(publicNetworkAccess) : (!empty(privateEndpoints) ? 'Disabled' : 'Enabled')
+    serviceManagedResourcesSettings: serviceManagedResourcesSettings
+    sharedPrivateLinkResources: sharedPrivateLinkResources // Note: This property is not idempotent. Neither with [] or `null`
```
Comment (AlexanderSehr) [inline avm/res/machine-learning-services/workspace/main.bicep:188]: I wonder if that comment should rather be in the parameter's description? 😄

File: avm/res/machine-learning-services/workspace/compute/main.bicep [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+
+resource machineLearningWorkspace 'Microsoft.MachineLearningServices/workspaces@2022-10-01' existing = {
+  name: machineLearningWorkspaceName
+}
+
+// ============ //
+// Deployments  //
+// ============ //
+
+resource machineLearningWorkspaceCompute 'Microsoft.MachineLearningServices/workspaces/computes@2022-10-01' = if (deployCompute == true) {
```
Comment (AlexanderSehr) [inline avm/res/machine-learning-services/workspace/compute/main.bicep:93]: I'd probably suggest to rename `machineLearningWorkspaceCompute` to just `compute` in alignment with how it's handled for other resources

File: avm/res/machine-learning-services/workspace/tests/e2e/defaults/main.test.bicep [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+  params: {
+    name: '${namePrefix}${serviceShort}001'
+    location: resourceLocation
+    associatedApplicationInsightsResourceId: nestedDependencies.outputs.applicationInsightsResourceId
+    associatedKeyVaultResourceId: nestedDependencies.outputs.keyVaultResourceId
+    associatedStorageAccountResourceId: nestedDependencies.outputs.storageAccountResourceId
+    sku: 'Basic'
+    managedIdentities: {
+      systemAssigned: true
+    }
```
Comment (AlexanderSehr) [inline avm/res/machine-learning-services/workspace/tests/e2e/defaults/main.test.bicep:61]:
```suggestion
```

Not mandatory (actually the value is the default of the parameter)

File: avm/res/machine-learning-services/workspace/tests/e2e/encr/dependencies.bicep [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+resource keyVault 'Microsoft.KeyVault/vaults@2022-07-01' = {
+  name: keyVaultName
+  location: location
+  properties: {
+    sku: {
+      family: 'A'
+      name: 'standard'
+    }
+    tenantId: tenant().tenantId
+    enablePurgeProtection: true // Required by batch account
```
Comment (AlexanderSehr) [inline avm/res/machine-learning-services/workspace/tests/e2e/encr/dependencies.bicep:52]:
```suggestion
    enablePurgeProtection: true // Required for CMK
```

File: avm/res/machine-learning-services/workspace/tests/e2e/encr/main.test.bicep [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+          nestedDependencies.outputs.privateDNSZoneResourceId
+        ]
+        tags: {
+          'hidden-title': 'This is visible in the resource name'
+          Environment: 'Non-Prod'
+          Role: 'DeploymentValidation'
+        }
+      }
+    ]
+    // systemAssigned must be false if `primaryUserAssignedIdentity` is provided
```
Comment (AlexanderSehr) [inline avm/res/machine-learning-services/workspace/tests/e2e/encr/main.test.bicep:85]:
```suggestion
```

File: avm/res/machine-learning-services/workspace/tests/e2e/encr/main.test.bicep [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+      systemAssigned: false
+      userAssignedResourceIds: [
+        nestedDependencies.outputs.managedIdentityResourceId
+      ]
+    }
+    tags: {
+      'hidden-title': 'This is visible in the resource name'
+      Environment: 'Non-Prod'
+      Role: 'DeploymentValidation'
+    }
```
Comment (AlexanderSehr) [inline avm/res/machine-learning-services/workspace/tests/e2e/encr/main.test.bicep:96]:
```suggestion
```

not required for CMK

File: avm/res/machine-learning-services/workspace/tests/e2e/encr/main.test.bicep [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+        privateDnsZoneResourceIds: [
+          nestedDependencies.outputs.privateDNSZoneResourceId
+        ]
+        tags: {
+          'hidden-title': 'This is visible in the resource name'
+          Environment: 'Non-Prod'
+          Role: 'DeploymentValidation'
+        }
+      }
+    ]
```
Comment (AlexanderSehr) [inline avm/res/machine-learning-services/workspace/tests/e2e/encr/main.test.bicep:84]:
```suggestion
```

not required for CMK

File: avm/res/machine-learning-services/workspace/tests/e2e/encr/main.test.bicep [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+resource resourceGroup 'Microsoft.Resources/resourceGroups@2021-04-01' = {
+  name: resourceGroupName
+  location: resourceLocation
+}
+
+module nestedDependencies 'dependencies.bicep' = {
+  scope: resourceGroup
+  name: '${uniqueString(deployment().name, resourceLocation)}-nestedDependencies'
+  params: {
+    virtualNetworkName: 'dep-${namePrefix}-vnet-${serviceShort}'
```
Comment (AlexanderSehr) [inline avm/res/machine-learning-services/workspace/tests/e2e/encr/main.test.bicep:41]: Can be removed from dependencies as not required for CMK

File: avm/res/machine-learning-services/workspace/tests/e2e/encr/main.test.bicep [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+        tags: {
+          'hidden-title': 'This is visible in the resource name'
+          Environment: 'Non-Prod'
+          Role: 'DeploymentValidation'
+        }
+      }
+    ]
+    // systemAssigned must be false if `primaryUserAssignedIdentity` is provided
+    managedIdentities: {
+      systemAssigned: false
```
Comment (AlexanderSehr) [inline avm/res/machine-learning-services/workspace/tests/e2e/encr/main.test.bicep:87]:
```suggestion
    managedIdentities: {
      // systemAssigned must be false if `primaryUserAssignedIdentity` is provided
      systemAssigned: false
```

File: avm/res/machine-learning-services/workspace/tests/e2e/max/main.test.bicep [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+            minNodeCount: 0
+            nodeIdleTimeBeforeScaleDown: 'PT5M'
+          }
+          vmPriority: 'Dedicated'
+          vmSize: 'STANDARD_DS11_V2'
+        }
+        sku: 'Basic'
+        // Must be false if `primaryUserAssignedIdentity` is provided
+        managedIdentities: {
+          systemAssigned: false
```
Comment (AlexanderSehr) [inline avm/res/machine-learning-services/workspace/tests/e2e/max/main.test.bicep:99]:
```suggestion
        managedIdentities: {
          systemAssigned: false // Must be false if `primaryUserAssignedIdentity` is provided
```

File: avm/res/machine-learning-services/workspace/main.bicep [hunk 14]
```diff
@@ -0,0 +1,477 @@
...
+      category: group.?category
+      enabled: group.?enabled ?? true
+    }]
+    marketplacePartnerId: diagnosticSetting.?marketplacePartnerResourceId
+    logAnalyticsDestinationType: diagnosticSetting.?logAnalyticsDestinationType
+  }
+  scope: workspace
+}]
+
+module workspace_privateEndpoints 'br/public:avm/res/network/private-endpoint:0.4.0' = [for (privateEndpoint, index) in (privateEndpoints ?? []): {
```
Comment (AlexanderSehr) [inline avm/res/machine-learning-services/workspace/main.bicep:246]: The [PE schema update](https://azure.github.io/Azure-Verified-Modules/specs/shared/interfaces/#private-endpoints) was merged. Would be great if we could apply it here.

File: avm/res/machine-learning-services/workspace/compute/main.bicep (hunk 5 above)
Comment (krbar) [reply avm/res/machine-learning-services/workspace/compute/main.bicep]: done: https://github.com/Azure/bicep-registry-modules/pull/1055/commits/31ca316848a9674042a411cb588468893f253f3e

File: avm/res/machine-learning-services/workspace/main.bicep (hunk 4 above)
Comment (krbar) [reply avm/res/machine-learning-services/workspace/main.bicep]: Moved it there: [86efb8b](https://github.com/Azure/bicep-registry-modules/pull/1055/commits/86efb8bf26f57ae8d6a923b595adea3a9b8e536c)

File: avm/res/machine-learning-services/workspace/tests/e2e/encr/main.test.bicep (hunk 11 above)
Comment (krbar) [reply avm/res/machine-learning-services/workspace/tests/e2e/encr/main.test.bicep]: done: https://github.com/Azure/bicep-registry-modules/pull/1055/commits/60a223209a4f8e70dca424c06b797119d9cdbe30

File: avm/res/machine-learning-services/workspace/main.bicep (hunk 14 above)
Comment (krbar) [reply avm/res/machine-learning-services/workspace/main.bicep]: Done: [9f771a9](https://github.com/Azure/bicep-registry-modules/pull/1055/commits/9f771a9462ed9130ac3f7130e6c7ddd0755437a9)

File: avm/res/machine-learning-services/workspace/tests/e2e/waf-aligned/main.test.bicep [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+        // Must be false if `primaryUserAssignedIdentity` is provided
+        managedIdentities: {
+          systemAssigned: false
+          userAssignedResourceIds: [
+            nestedDependencies.outputs.managedIdentityResourceId
+          ]
+        }
+      }
+    ]
+    description: 'The cake is a lie.'
```
[the remark this answers is not in the document]
Comment (krbar) [reply avm/res/machine-learning-services/workspace/tests/e2e/waf-aligned/main.test.bicep]: I reduced the params in `waf-aligned` accordingly.

**Comment by krbar** [comment]
@AlexanderSehr 
I removed the parameter `sharedPrivateLinkResources` from the resource properties if it's empty to enable idempotency.

https://github.com/Azure/bicep-registry-modules/pull/1055/commits/0fe14d1bcf83d9721038a490784f1d37ff0be9f8
https://github.com/Azure/bicep-registry-modules/pull/1055/commits/dd1a523695459d4931181acd2cf450868162eb03

File: avm/res/machine-learning-services/workspace/main.bicep [hunk 16]
```diff
@@ -243,21 +249,33 @@ resource workspace_diagnosticSettings 'Microsoft.Insights/diagnosticSettings@202
...
-    ]
-    name: privateEndpoint.?name ?? 'pep-${last(split(workspace.id, '/'))}-${privateEndpoint.?service ?? 'amlworkspace'}-${index}'
+    ] : null
+    manualPrivateLinkServiceConnections: privateEndpoint.?manualPrivateLinkServiceConnections == true ? [
+      {
+        name: privateEndpoint.?privateLinkServiceConnectionName ?? '${last(split(workspace.id, '/'))}-${privateEndpoint.?service ?? 'amlworkspace'}-${index}'
+        properties: {
+          privateLinkServiceId: workspace.id
+          groupIds: [
+            privateEndpoint.?service ?? ''
```
Comment (AlexanderSehr) [inline avm/res/machine-learning-services/workspace/main.bicep:273]:
```suggestion
            privateEndpoint.?service ?? 'amlworkspace'
```

File: avm/res/machine-learning-services/workspace/main.bicep [hunk 17]
```diff
@@ -243,21 +249,33 @@ resource workspace_diagnosticSettings 'Microsoft.Insights/diagnosticSettings@202
...
-module workspace_privateEndpoints 'br/public:avm/res/network/private-endpoint:0.4.0' = [for (privateEndpoint, index) in (privateEndpoints ?? []): {
+module workspace_privateEndpoints 'br/public:avm/res/network/private-endpoint:0.4.1' = [for (privateEndpoint, index) in (privateEndpoints ?? []): {
   name: '${uniqueString(deployment().name, location)}-workspace-PrivateEndpoint-${index}'
   params: {
-    privateLinkServiceConnections: [
+    name: privateEndpoint.?name ?? 'pep-${last(split(workspace.id, '/'))}-${privateEndpoint.?service ?? 'amlworkspace'}-${index}'
+    privateLinkServiceConnections: privateEndpoint.?manualPrivateLinkServiceConnections != true ? [
       {
-        name: name
+        name: privateEndpoint.?privateLinkServiceConnectionName ?? '${last(split(workspace.id, '/'))}-${privateEndpoint.?service ?? ''}-${index}'
```
Comment (AlexanderSehr) [inline avm/res/machine-learning-services/workspace/main.bicep:258]:
```suggestion
        name: privateEndpoint.?privateLinkServiceConnectionName ?? '${last(split(workspace.id, '/'))}-${privateEndpoint.?service ?? 'amlworkspace'}-${index}'
```

File: avm/res/machine-learning-services/workspace/tests/e2e/max/main.test.bicep [hunk 18]
```diff
@@ -0,0 +1,170 @@
...
+        subnetResourceId: nestedDependencies.outputs.subnetResourceId
+        privateDnsZoneResourceIds: [
+          nestedDependencies.outputs.privateDNSZoneResourceId
+        ]
+        tags: {
+          'hidden-title': 'This is visible in the resource name'
+          Environment: 'Non-Prod'
+          Role: 'DeploymentValidation'
+        }
+      }
```
[the remark this answers is not in the document]
Comment (krbar) [reply avm/res/machine-learning-services/workspace/tests/e2e/max/main.test.bicep:139]: Changes applied, and re-tested.

**Review by AlexanderSehr — APPROVED** [review]

**Pull request merged**